### DISCUSSION TITLE
Waku Islands?

### DISCUSSION THREAD

**Discussion by tylersayshi** [wakujs/waku#816] [Ideas]
Would it be useful to the project for someone to write a guide for how to essentially achieve "astro islands" with waku?

Their explanation is here: https://astro.build/blog/future-of-astro-server-islands/
And docs here: https://docs.astro.build/en/concepts/islands/

My understanding of how they pose optimizing their sites is by taking the static pre-rendered version of their site, put that on a cdn... then embed js in the static build to stream in "what is left" from the server.

Given the current state of waku, where it can compose this behavior and streaming easily with RSC's, my initial guess is that it would at least be a catchy walkthrough for new users who appreciate astro, but want to stay fully in a react project. 

If this is something we want to support and give docs for how to do, it might simultaneously be a good experiment for how to deploy a slightly more complex version than what is currently being proposed. This could lead to some good edge-case finding and whatnot.

**Comment by tylersayshi** [comment]
Actually... is this just what the vercel build is doing already with splitting the output between a cdn and the node server?

**Comment by dai-shi** [comment]
(I'm not 100% sure if I understand the server islands and what you actually expect, but:)

RSC is technically capable of generating static shell/skeleton contents in advance and streaming dynamic contents on demand (with or without SSR), using `<Suspense>`.
Currently, waku/router only supports page-based static/dynamic division, and it's limited. (Using Minimal API, it's less limited, but still probably not fully supported.)

I am _very_ interested in supporting component-based static/dynamic division, but that would require deeper understanding and more experiments. So, my plan for now is to tackle it after v1 or maybe v1-beta. I may change my mind though.

> Given the current state of waku, where it can compose this behavior and streaming easily with RSC's

Maybe you are referring to something else than what I described above.

**Reply by dai-shi** [reply]
Just watching @t3dotgg video, and yeah, it seems "server islands" look closer to what I would like to accomplish _with_ RSC.

<img width="628" alt="image" src="https://github.com/user-attachments/assets/992239eb-c0b5-47ba-951e-d42df37b9181">

**Reply by tylersayshi** [reply]
Awesome! That video was exactly where I took saw the similarity to what waku can be capable of as well. 

RSC and astro are coming from opposite directions, but the solution is very similar (in simple terms): Pre-build a "shell" on the  CDN and then specify _parts_ of the UI that need to be interactive. 

Astro does this by adding in "islands" for interactivity. RSC gets to this by enabling PPR that can be stored on a CDN.

**Reply by dai-shi** [reply]
As @t3dotgg points out, I think Nextjs's PPR is too complicated and may require cloud integration, so my plan is not to take that road. We may still call it PPR though.

**Reply by tylersayshi** [reply]
That makes a ton of sense and fwiw I completely agree :)

**Comment by dai-shi** [comment]
I started experimenting something: #1166

**Comment by dai-shi** [comment]
We'll be releasing lazy slices, which will cover the "Waku Islands" use case.

**Comment by dai-shi** [comment]
v0.25.0 is out.
